app/server/lib/AttachmentFileManager.ts

```diff
@@ -108,24 +144,253 @@ export class AttachmentFileManager implements IAttachmentFileManager {
+    }
+    return Promise.resolve();
+  }
+
+  public transferStatus() {
+    return {
+      pendingTransferCount: this._pendingFileTransfers.size,
+      isRunning: this._transferJob && !this._transferJob.isFinished || false
+    };
+  }
+
+  private _runTransferJob(): TransferJob {
+    if (this._transferJob && this.transferStatus().isRunning) {
+      return this._transferJob;
+    }
+    const transferPromise = this._performPendingTransfers();
+    const newTransferJob: TransferJob = {
+      isFinished: false,
+      promise: transferPromise,
+    };
+
+    newTransferJob.promise.finally(() => newTransferJob.isFinished = true);
+
+    this._transferJob = newTransferJob;
+
+    return newTransferJob;
+  }
+
+  private async _performPendingTransfers() {
+    while (this._pendingFileTransfers.size > 0) {
+      // Map.entries() will always return the most recent key/value from the map, even after a long async delay
+      // Meaning we can safely iterate here and know the transfer is up to date.
+      for (const [fileIdent, targetStoreId] of this._pendingFileTransfers.entries()) {
+        try {
+          await this.transferFileToOtherStore(fileIdent, targetStoreId);
+        } catch(e) {
+          this._log.warn({ fileIdent, storeId: targetStoreId }, `transfer failed: ${e.message}`);
+        }
+        finally {
+          // If a transfer request comes in mid-transfer, it will need re-running.
+          if (this._pendingFileTransfers.get(fileIdent) === targetStoreId) {
+            this._pendingFileTransfers.delete(fileIdent);
+          }
+        }
+      }
+    }
+  }
+
+  private async _addFileToLocalStorage(
     fileIdent: string,
     fileData: Buffer
   ): Promise<AddFileResult> {
-    this._log.info({ fileIdent, storeId }, `adding file to ${storeId ? "external" : "document"} storage`);
-    if (storeId === undefined) {
-      return this._addFileToLocalStorage(fileIdent, fileData);
+      this._log.info({
+        fileIdent,
+      }, `adding file to document storage`);
+
+    const fileInfoNoData = await this._docStorage.getFileInfoNoData(fileIdent);
+    const fileExists = fileInfoNoData !== null;
+
+    if (fileExists) {
+      const isFileInLocalStorage = fileInfoNoData.storageId === null;
+      // File is already stored in a different store, the 'add file' operation shouldn't change that.
+      // One way this could happen, is if the store has changed and no migration has happened.
+      if (!isFileInLocalStorage) {
+        return {
+          fileIdent,
+          isNewFile: false,
+        };
+      }
     }
-    const store = await this._getStore(storeId);
-    if (!store) {
-      this._log.info({ fileIdent, storeId }, "tried to fetch attachment from an unavailable store");
-      throw new StoreNotAvailableError(storeId);
+
+    // A race condition can occur here, if the file's database record is modified between the
+    // `getFileInfoNoData` call earlier in this function and now.
```

Comment:
`ActiveDoc` has a well worked out mechanism to avoid parallel changes to the document (see the `Share` class). The underlying SQLite database also applies changes sequentially. It feels like something has gone wrong somewhere, there is some parallelism introduced for attachments, that is making your life harder.

Reply:
This is theoretical for the most part - I think it might be possible if several add requests for the file come in at once, and we don't apply the share class to the attachment API in ActiveDoc? 

I could probably remove this without any harm, but I thought it might be worth pointing out just in case in comes at some point in the future. Could also add "theoretical" to the start of the comment. Not sure what the best practice here is.

Reply:
Hmm the user actions applied in ActiveDoc.addAttachments should happen via the Share class ultimately. I feel that the Promise.all in there is silly. But it does up the odds of tickling race conditions like this that might be hard to spot otherwise, so that's good.

Ok, assuming this was the only race, fine to leave as is.